Adding host name and IP address guidelines (#28378)
openshift/openshift-docs: contributing_to_docs/doc_guidelines.adoc

$ oc get nodes
----
+
.Example output
[source,terminal]
----
NAME                          STATUS   ROLES    AGE   VERSION
compute-1.example.com         Ready    worker   33m   v1.19.0+9f84db3
control-plane-1.example.com   Ready    master   41m   v1.19.0+9f84db3
control-plane-2.example.com   Ready    master   45m   v1.19.0+9f84db3
compute-2.example.com         Ready    worker   38m   v1.19.0+9f84db3
compute-3.example.com         Ready    worker   33m   v1.19.0+9f84db3
control-plane-3.example.com   Ready    master   41m   v1.19.0+9f84db3
----
....

[NOTE]
====
Some provider-formatted host names include IPv4 addresses. An OpenShift Container Platform node name typically reflects the host name of a node. If node names in your output need to be provider-specific and require this format, use IPv4 addresses in the 192.0.2.0 to 192.0.2.255 reserved address range. For example, you could use `ip-192-0-2-1.example.com` as a node name that includes an IPv4 address.
====

== IPv4 address formatting

Where possible, use the default OpenShift Container Platform private IPv4 addresses in example output. Private IPv4 addresses are those that are included in the following reserved address ranges:

> * 10.0.0.0 to 10.255.255.255

Reviewer: For this, I'd include 10/8, 172.16/12, 192.168/16.

Reviewer: And perhaps class A, B, C as well.

Author: @jboxman thank you for these suggestions. I have now updated the list.

control-plane-1.example.com   Ready    master   41m   v1.19.0+9f84db3
control-plane-2.example.com   Ready    master   45m   v1.19.0+9f84db3
compute-2.example.com         Ready    worker   38m   v1.19.0+9f84db3
compute-3.example.com         Ready    worker   33m   v1.19.0+9f84db3
control-plane-3.example.com   Ready    master   41m   v1.19.0+9f84db3
----
....

[NOTE]
====
Some provider-formatted host names include IPv4 addresses. An OpenShift Container Platform node name typically reflects the host name of a node. If node names in your output need to be provider-specific and require this format, use private IPv4 addresses. For example, you could use `ip-10-0-48-9.example.com` as a node name that includes a private IPv4 address.
====

== IP addresses

You may include IPv4 addresses from test clusters in examples in the documentation, as long as they are private. Private IPv4 addresses fall into one of the following ranges:

* 10.0.0.0 to 10.255.255.255 (class A address block 10.0.0.0/8)
* 172.16.0.0 to 172.31.255.255 (class B address block 172.16.0.0/12)
* 192.168.0.0 to 192.168.255.255 (class C address block 192.168.0.0/16)

> You can replace public IPv4 addresses with addresses in the TEST-NET-3 address block reserved for documentation, which ranges from 203.0.113.0 to 203.0.113.255.  

Reviewer 2: "Can" or "must?"

Author: @maxwelldb thanks for highlighting this. I have updated the sentence to state "must".

Reviewer 2: Thanks. 

Again, I'm absolutely against adding a restriction to the ISG--yet another thing for writers and reviewers to look out for--for no clear benefit.

|You can use this format if you need to describe the bootstrap node role type within a node name.
|===

This example lists the status of cluster nodes that use the node name formatting guidelines:

....
[source,terminal]
----
$ oc get nodes
----
+
.Example output
[source,terminal]
----
NAME                          STATUS   ROLES    AGE   VERSION
compute-1.example.com         Ready    worker   33m   v1.19.0+9f84db3
control-plane-1.example.com   Ready    master   41m   v1.19.0+9f84db3
control-plane-2.example.com   Ready    master   45m   v1.19.0+9f84db3
compute-2.example.com         Ready    worker   38m   v1.19.0+9f84db3
compute-3.example.com         Ready    worker   33m   v1.19.0+9f84db3
control-plane-3.example.com   Ready    master   41m   v1.19.0+9f84db3
----
....

[NOTE]
====
Some provider-formatted host names include IPv4 addresses. An OpenShift Container Platform node name typically reflects the host name of a node. If node names in your output need to be provider-specific and require this format, use private IPv4 addresses. For example, you could use `ip-10-0-48-9.example.com` as a node name that includes a private IPv4 address.
====

== IPv4 address formatting

> Do not use public IPv4 addresses in commands or example output in the product documentation. If you include the output from a command run on an OpenShift Container Platform test cluster in the documentation, replace any public or sensitive IPv4 addresses. You can replace public IPv4 addresses with addresses in the 203.0.113.0/24 TEST-NET-3 address block, which range from 203.0.113.0 to 203.0.113.255. The TEST-NET-3 address block is reserved for documentation. See link:https://tools.ietf.org/html/rfc5737[IETF RFC 5737 - IPv4 Address Blocks Reserved for Documentation] for more information about IPv4 address ranges that are reserved for documentation.

Reviewer 3: Why aren't we using the IP address range that the style guide requires?

> * 192.0.2.0/24 (or 192.0.2.*)
> * 198.51.100.0/24 (or 198.51.100.*)
> * 203.0.113.0/24 (or 203.0.113.*)

Reviewer 2: +1. Not sure why we're deviating from the ranges set in the style guide.

Reviewer 4: I might be missing something here, but it looks like `203.0.113.0/24` **is** the range mentioned in the style guide (3rd bullet). The rest of the sentence (`which range from 203.0.113.0 to 203.0.113.255`) is an explanation. The explanation might be overkill, however, because I would expect a contributor to be familiar with CIDR notation.

Author: Thank you for your input!

During a discussion last week, @mikemckiernan highlighted that the 192.0.2.0/24 TEST-NET-1 and 198.51.100.0/24 TEST-NET-2 documentation addresses might look too similar to the 192.168.0.0/16 private address range. Since we are suggesting reserved documentation addresses as a replacement for public IPs, we put forward a recommendation to use the 203.0.113.0/24 TEST-NET-3 documentation address block specifically. That block does not look similar to the private address range.

Also, this PR initially suggested using the TEST-NET addresses for example output. @mikemckiernan highlighted that it would be better to use the private IP addresses that are output by OpenShift by default.

So, the current proposal essentially highlights that it is ok to use the private IP addresses included in output from test clusters, but provides guidance on how to replace public IPs, since they might resolve to a real world node.

Please let me know if you would like me to add further clarification within the content about this.

Reviewer 5: The private ranges are larger than the default cluster network and default service network, 10.128.0.0/14 and 172.30.0.0/16, respectively. The clusters that associates access use the default values and my feeling is that we want to repeat them in product documentation to reinforce customer understanding.

The use of the 203.0.113.0/24 range to replace public IP addresses provides good visual distinction from the private ranges.  My expectation is customers can quickly scan sample output, see a 203 range IP address, and recognize that their output should differ--according to the cloud provider their cluster uses.

Reviewer 2: I haven't seen any indications that this is taxing on customers, but it would generate some additional work for us. 

I'd prefer to keep the ISG convention. The change seems like it could increase cognitive load for writers to address something that, so far, hasn't popped up as a customer concern. 

Is the proposal that this rule would be a "going forward" thing, or would every actively developed version of the docs be updated upon its implementation?

Author: @maxwelldb thank you for your comment. Can you please let me know which part of the ISG convention you are referring to? I want to ensure that I understand your preference here properly. Thanks!

Reviewer 2: @pneedle-rh The IPv4 section in "Chapter 5 - References" -> "References to web addresses, protocols, and IP addresses."

Author: @maxwelldb so is my understanding correct then that your preference is for the following, from the **IPv4 addresses** subsection in that chapter?:

"If your example already contains the loopback address and you need additional example
IPv4 addresses, use addresses in the /8 IP address block that are reserved for Test-Net: 192.0.2.0
through 192.0.2.24."

Reviewer 3: @pneedle-rh, the new version of the style guide does not mention the Test-Net ranges.

~~~
You can also use the IP loopback address, which is 127.0.0.0/8 (127.*.*.*), although typically, only 127.0.0.1 is used for loopback purposes.
You can include wildcard characters in examples of IPv4 addresses to illustrate that a range of values is permitted; for example, 127.*.*.12 includes all addresses that start with 127 and end with 12.
If you use wildcards, do not use identifiable companies' addresses in documentation. For example, using 192.*.*.12 is permissible because the 192.0.0.0/8 range covers a variety of uses.
~~~

Reviewer 2: @pneedle-rh Oh, I was using the paper guide. In the newest version (p. 258 of the PDF that's in circulation): 

> If you must use an IP address in an example, use an address that is reserved for examples, and not one that relates to a particular company, unless that is required. For example, all addresses in the range 9.0.0.0/8 (or 9.\*.\*.\*) are owned by IBM. Three ranges of addresses are reserved for use in documentation:
> 
> * 192.0.2.0/24 (or 192.0.2.*)
> * 198.51.100.0/24 (or 198.51.100.*)
> * 203.0.113.0/24 (or 203.0.113.*)
> 
> ... 
> 
> You can also use the IP loopback address, which is 127.0.0.0/8 (127.\*.\*.\*), although typically, only 127.0.0.1 is used for loopback purposes. 
> 
> You can include wildcare characters in examples of IPv4 addresses to illustrate that a range of values is permittedl; for example, 127.\*.\*.12 includes all addresses that start with 127 and end with 12. If you use wildcards, do not use identifiable companies' addresses in documentation. For example, using 192.\*.\*.12 is permissible because the 192.0.0.0/8 range covers a variety of uses.

Author: @kalexand-rh and @maxwelldb thank you. It looks like the version of the ISG I have is out of date. I will update my copy.

The 192.0.2.0/24, 198.51.100.0/24 and 203.0.113.0/24 that are referenced in Max's updated ISG example are the three TEST-NET ranges, as detailed in https://tools.ietf.org/html/rfc5737.

Thank you both for clarifying your preference for the guidelines.

control-plane-1.example.com   Ready    master   41m   v1.19.0+9f84db3
control-plane-2.example.com   Ready    master   45m   v1.19.0+9f84db3
compute-2.example.com         Ready    worker   38m   v1.19.0+9f84db3
compute-3.example.com         Ready    worker   33m   v1.19.0+9f84db3
control-plane-3.example.com   Ready    master   41m   v1.19.0+9f84db3
----
....

[NOTE]
====
Some provider-formatted host names include IPv4 addresses. An OpenShift Container Platform node name typically reflects the host name of a node. If node names in your output need to be provider-specific and require this format, use private IPv4 addresses. For example, you could use `ip-10-0-48-9.example.com` as a node name that includes a private IPv4 address.
====

== IP addresses

You may include IPv4 addresses from test clusters in examples in the documentation, as long as they are private. Private IPv4 addresses fall into one of the following ranges:

* 10.0.0.0 to 10.255.255.255 (class A address block 10.0.0.0/8)
* 172.16.0.0 to 172.31.255.255 (class B address block 172.16.0.0/12)
* 192.168.0.0 to 192.168.255.255 (class C address block 192.168.0.0/16)

> If you have a public IP address that you need to replace, use an address in one of the following address blocks reserved for documentation:

Reviewer 2: ```suggestion
Replace all public IP addresses with an address from the following blocks that are reserved for documentation:
```

Reviewer 2: Or end at "blocks:" and tack "These address blocks are reserved for documentation." after the UL.

Author: Thank you for the suggestion. I have now applied this.

Do not use markup in headings.

> == Host name formatting

Reviewer 6: Although the node name might typically be the host name, it might not be immediately obvious to me (or maybe new people) that this applies to the node name. And maybe potentially other areas of the objects.

I'd suggest to give some examples of where this could be used, like the `oc get nodes` output we've been talking about.

Author: Thank you for highlighting this. I have updated most of the "host name" references to "node name", except where referring to provider host names.

I have also added an example node name output to demonstrate usage.

Do not use internal company server names in commands or example output. Provide generic host name examples that are not provider-specific, unless required. Where possible, use the example.com domain name when providing fully qualified domain names (FQDNs).

The following table includes example OpenShift Container Platform 4 host names and their corresponding role types:

[options="header"]
|===

|Host name |Role type

|*node1.example.com*
.3+.^|You can use this format for hosts that do not need role-specific host names.

|*node2.example.com*

|*node3.example.com*

|*control-plane-1.example.com*
.3+.^|You can use this format if you need to describe the control plane role type within a host name.

> |*control-plane-2.example.com*

Author: @bergerhoffer thank you for your review! I have now changed the first three node name examples to `node-1.example.com`, `node-2.example.com` and `node-3.example.com`, for consistency.
